### ISSUE TITLE
Unit tests in a package do not show up in Rider's Unit Tests window

### ISSUE THREAD

**Issue by sitterheim** [JetBrains/resharper-unity#2344]
I created a package with tests:
![image](https://user-images.githubusercontent.com/42713966/184533579-82ee2e80-0961-4239-abfb-c1fa729c357d.png)

Actually, I have several packages with tests, this is just an example.

Rider Unit Tests window only shows the tests that are part of the project itself, ie located under "Assets":
![image](https://user-images.githubusercontent.com/42713966/184533620-a6e9c07b-9cf5-494b-bb57-1545664ac5f2.png)

The packages in question all have `"hideInEditor": false` in their package.json. Consequently, Test Runner in the editor shows all the tests from these packages:

![image](https://user-images.githubusercontent.com/42713966/184533663-f1dd25c1-4eeb-494f-99ec-79d2471c3865.png)

Observed with these versions:
Unity 2022.1.13f1
JetBrains Rider Editor 3.0.15

**Comment by van800** [comment]
Try to open some file with tests, would they get detected? You may also try to build project with tests or the whole solution in Rider. Would that help?

**Comment by sitterheim** [comment]
The MeshGeometryTests (located under Assets) do show up.
I did a "Rebuild Solution" that completed without errors, but Unit Tests window contents did not change (shows only tests located under Assets).

**Comment by van800** [comment]
I would be interested to look in the logs. Call Help - Report a Bug, agree to attach logs. Post the link here. Thanks!

**Comment by sitterheim** [comment]
You mean this link?
https://dotnet-ea.services.jetbrains.com/files/large/1a42af1d-0a6e-4507-9ac6-d437b974774e/rider-logs-attachment34834.tar.xz

Should I submit the issue on youtrack as well?

**Comment by citizenmatt** [comment]
Is this related to the `testables` element in `Packages/manifest.json`? I think the package name needs to be added to this array. And it's worth checking that [`optionalUnityReferences` in the `.asmdef` file](https://docs.unity3d.com/Manual/cus-tests.html) includes `TestAssemblies`.

Also, could you clarify what `hideInEditor` does? Does this show/hide the package in the Project view?

**Comment by van800** [comment]
JFYI https://youtrack.jetbrains.com/issue/RIDER-81034/Unity-TestRunner-tests-for-visible-packages-are-missing-from-Riders-Unit-Tests-window

**Comment by sitterheim** [comment]
`hideInEditor` makes package assets appear (ie you can select materials, meshes, etc. located in a package) and *I think* it makes package tests appear in TestRunner. At least it wasn't necessary for the particular package I work with, its tests are already in TestRunner.

Nevertheless, I added this package to the project's manifest under testables, restarted Unity and Rider, but still none of the package tests show up in Rider's Unity Tests window.

**Comment by sitterheim** [comment]
Note: adding the "TestAssemblies" as optional references simply duplicates the UnityEngine/Editor.TestRunner references.

**Comment by van800** [comment]
@sitterheim 
Please share the `Packages/manifest.json` file here or in the YT issue.
Do you know to which file the `hideInTheEditor` is written
I have quickly checked our project https://github.com/van800/com.unity.ide.rider
and tests in the local package are found just fine with versions of Rider and Unity similar to yours.

**Comment by sitterheim** [comment]
It's a lot :)

```
{
  "dependencies": {
    "com.codesmile.core": "file:C:/Users/Eedee F/UnityProjects/_MyPackages/CodeSmile-Package",
    "com.codesmile.devassets": "file:C:/Users/Eedee F/UnityProjects/_MyPackages/CodeSmile-DevAssetsPackage",
    "com.codesmile.gmesh": "file:C:/Users/Eedee F/UnityProjects/_MyPackages/GMesh",
    "com.codesmile.node-graph-processor": "file:C:/Users/Eedee F/UnityProjects/_MyPackages/NodeGraphProcessor/Assets/Plugins/NodeGraphProcessor",
    "com.codesmile.voxelheart": "file:C:/Users/Eedee F/UnityProjects/_MyPackages/VoxelHeart/VoxelHeart-Package",
    "com.codesmile.voxelheart-devassets": "file:C:/Users/Eedee F/UnityProjects/_MyPackages/VoxelHeart/VoxelHeart-DevAssetsPackage",
    "com.exppad.bmesh-unity": "file:C:/Users/Eedee F/UnityProjects/_MyPackages/BMeshUnity",
    "com.jacksondunstan.native-collections": "file:C:/Users/Eedee F/UnityProjects/_MyPackages/NativeCollections",
    "com.unity.ide.rider": "3.0.15",
    "com.unity.jobs": "0.70.0-preview.7",
    "com.unity.performance.profile-analyzer": "1.1.1",
    "com.unity.recorder": "3.0.3",
    "com.unity.render-pipelines.universal": "13.1.8",
    "com.unity.test-framework": "1.1.33",
    "com.unity.test-framework.performance": "2.8.1-preview",
    "com.unity.textmeshpro": "3.0.6",
    "com.unity.ugui": "1.0.0",
    "com.unity.modules.ai": "1.0.0",
    "com.unity.modules.androidjni": "1.0.0",
    "com.unity.modules.animation": "1.0.0",
    "com.unity.modules.assetbundle": "1.0.0",
    "com.unity.modules.audio": "1.0.0",
    "com.unity.modules.cloth": "1.0.0",
    "com.unity.modules.director": "1.0.0",
    "com.unity.modules.imageconversion": "1.0.0",
    "com.unity.modules.imgui": "1.0.0",
    "com.unity.modules.jsonserialize": "1.0.0",
    "com.unity.modules.particlesystem": "1.0.0",
    "com.unity.modules.physics": "1.0.0",
    "com.unity.modules.physics2d": "1.0.0",
    "com.unity.modules.screencapture": "1.0.0",
    "com.unity.modules.terrain": "1.0.0",
    "com.unity.modules.terrainphysics": "1.0.0",
    "com.unity.modules.tilemap": "1.0.0",
    "com.unity.modules.ui": "1.0.0",
    "com.unity.modules.uielements": "1.0.0",
    "com.unity.modules.umbra": "1.0.0",
    "com.unity.modules.unityanalytics": "1.0.0",
    "com.unity.modules.unitywebrequest": "1.0.0",
    "com.unity.modules.unitywebrequestassetbundle": "1.0.0",
    "com.unity.modules.unitywebrequestaudio": "1.0.0",
    "com.unity.modules.unitywebrequesttexture": "1.0.0",
    "com.unity.modules.unitywebrequestwww": "1.0.0",
    "com.unity.modules.vehicles": "1.0.0",
    "com.unity.modules.video": "1.0.0",
    "com.unity.modules.vr": "1.0.0",
    "com.unity.modules.wind": "1.0.0",
    "com.unity.modules.xr": "1.0.0"
  },
  "testables": [
    "com.codesmile.core",
	"com.codesmile.gmesh",
    "com.codesmile.voxelheart",
	"com.codesmile.node-graph-processor",
	"com.jacksondunstan.native-collections"
  ]
}

```

**Comment by sitterheim** [comment]
This is the package.json for GMesh:

```
{
  "name": "com.codesmile.gmesh",
  "version": "0.1.0",
  "displayName": "CodeSmile GMesh",
  "description": "GMesh enables creating reliable geometry editing tools with Unity Job System support.",
  "category": "Tools",
  "unity": "2020.3",
  "dependencies": {
    "com.unity.burst": "1.7.3",
    "com.unity.collections": "1.4.0",
    "com.unity.jobs": "0.70.0-preview.7",
    "com.unity.mathematics": "1.2.6",
    "com.unity.test-framework.performance": "2.8.1-preview"
  },
  "relatedPackages": {},
  "samples": [
    {
      "displayName": "GMesh Examples",
      "description": "Some examples ...",
      "path": "Samples~/Scenes"
    }
  ],
  "keywords": [
    "mesh",
    "model",
    "graph",
    "edit",
    "asset",
    "tool",
    "editor"
  ],
  "author": {
    "name": "CodeSmile",
    "email": "removed@gmail.com",
    "url": "https://forum.unity.com/forums/made-with-unity.11/"
  },
  "hideInEditor": false
}
```

**Comment by van800** [comment]
Please also show the asmdef from GMesh.

**Comment by sitterheim** [comment]
Certainly ...

CodeSmile.GMesh.asmdef
```
{
    "name": "CodeSmile.GMesh",
    "rootNamespace": "",
    "references": [
        "GUID:343deaaf83e0cee4ca978e7df0b80d21",
        "GUID:2bafac87e7f4b9b418d9448d219b01ab",
        "GUID:e0cd26848372d4e5c891c569017e11f1",
        "GUID:d8b63aba1907145bea998dd612889d6b",
        "GUID:2665a8d13d1b3f18800f46e256720795"
    ],
    "includePlatforms": [],
    "excludePlatforms": [],
    "allowUnsafeCode": false,
    "overrideReferences": false,
    "precompiledReferences": [],
    "autoReferenced": true,
    "defineConstraints": [],
    "versionDefines": [],
    "noEngineReferences": false
}
```
Since it uses GUIDs here are the friendly names of the references:
![image](https://user-images.githubusercontent.com/42713966/185350904-804755c8-6add-45b7-8f7e-dc8ef4c017cc.png)


CodeSmile.GMesh.Tests.Editor.asmdef
```
{
    "name": "CodeSmile.GMesh.Tests.Editor",
    "rootNamespace": "",
    "references": [
        "UnityEngine.TestRunner",
        "UnityEditor.TestRunner",
        "CodeSmile.GMesh",
        "Unity.Collections",
        "Unity.Mathematics",
        "BMeshUnity"
    ],
    "includePlatforms": [
        "Editor"
    ],
    "excludePlatforms": [],
    "allowUnsafeCode": false,
    "overrideReferences": true,
    "precompiledReferences": [
        "nunit.framework.dll"
    ],
    "autoReferenced": false,
    "defineConstraints": [],
    "versionDefines": [],
    "noEngineReferences": false
}
```

After all this, I wouldn't be surprised if it came down to some obvious setting I need to change/add/remove. ;)

**Comment by sitterheim** [comment]
Btw, the GMesh package is on github in case you need to look into other files:
https://github.com/sitterheim/GMesh/tree/develop

**Comment by sitterheim** [comment]
I looked into the Rider package.json and noticed this:

```
  "relatedPackages": {
    "com.unity.ide.rider.tests": "3.0.15"
  },
```

That seems to indicate that Rider tests are in another (tests-only?) package.

Is there any reason to move tests into a separate package beyond avoiding to distribute them to end users?

**Comment by van800** [comment]
You are right, we don't want to distribute tests and their additional dependencies with the package.

**Comment by van800** [comment]
Still not sure, what is wrong on your side. I have tried to reproduce the problem with the package from the github and it works.
<img width="1098" alt="image" src="https://user-images.githubusercontent.com/1482681/185458210-c9d4aa93-d0aa-4a14-9b11-1d41bde5b81f.png">


Could you please open Solution View instead of Unity view in Rider and check if you have projects similar to mine on the screenshot?

**Comment by van800** [comment]
Please, also share the `CodeSmile.GMesh.Tests.Editor.csproj` file. Thanks!

**Comment by sitterheim** [comment]
Thanks for sharing this screenshot! When I saw it I knew something was different on my end. Besides that I have it on the right pane.

This is where I don't see the tests:
![image](https://user-images.githubusercontent.com/42713966/185482698-626667c5-e2b2-49e7-896b-959288532eff.png)

For some reason there's a "Session" tab.
I switched over to "Explorer" and --- yeah, that's where all the tests are! :(

![image](https://user-images.githubusercontent.com/42713966/185482938-c52ea426-a0c1-45c4-8341-89dd96632330.png)

Sorry for all the effort you spent on me just having to switch tabs. OMG can't believe it.
